### ISSUE TITLE
Allow arbitrary folders to be synchronized

### ISSUE THREAD

**Issue by ThomasSteinbach** [nextcloud/android#21]
[enhancement]

Currently the nextCloud app will create a one to one folder representation of the remote server when synchronizing a directory. However it would be really helpful on a mobile device to map a device directory to a remote server directory.

A use case would be syncronizing a workspace directory of an app on the mobile device with a dedicated directory path on the remote server. For instance a directory with pictures often edited by a mobile app synchronized with a subpath of 'Photos' on the remote server.

This would assume issue #19 to be working, such that changes on the mobile device are propagated to the remote server.

**Comment by AndyScherzinger** [comment]
Afaik also a work in progress, by @tobiasKaminsky 
Branch can be found here: https://github.com/nextcloud/android/tree/extendInstantUpload

It probably isn't a 100% fit to your description but this is what we try to achieve on the branch:
>This allows to choose any folder on the phone to sync in one direction to owncloud, just like the instant upload does this for images.
This can fix the problem of not using the correct NEW_PHOTO Action and it can also help to "backup" important folders (like WhatsApp Images).


So this issue is probably an extension to the branch as to do the same thing but in a 2-way-sync manor.

**Comment by ThomasSteinbach** [comment]
Would there be a persistent relationship between the 'instantly uploaded folder' on the mobile device and the target directory on the nextCloud server? Meaning that if I instatly upload the folder again, the app just consider the files that have changed since the last upload.

**Comment by tobiasKaminsky** [comment]
I am not sure if the android permission system allows this.
Imagine we take whatsapp folder to synchronise 2-way to server/whatsapp. If we delete files on the server this means that the files are also deleted in whatsapp folder by our app.

**Comment by AvonSyr** [comment]
It should be work. Currently I'm using the Andoird app "FolderSync" to 2-way-sync multiple directories between Android and Nextcloud. Would be awesome if this functionality would be provided by the Nextcloud app itself.

**Comment by codethief** [comment]
I second @ThomasSteinbach's request. Along with proper 2-way sync, this is the #1 feature I am waiting for.

**Comment by spacemanspiff2007** [comment]
+1

**Comment by Micha-Btz** [comment]
@tobiasKaminsky 
is it included in the nextcloud beta app in f-droid store or must i compile this branch by myself?

**Comment by AndyScherzinger** [comment]
@Micha-Btz This is not implemented yet and the started implementation has to be changed too due to the Android N changes, so the mentioned implementations isn't available yet. So Yes you would have to compile the branch yourself at this moment. Sorry :(

**Comment by tobiasKaminsky** [comment]
Although we made great progress with the latest beta, as I understand we still need a custom folder for auto upload. @AndyScherzinger this is something you have on your roadmap, right?

**Comment by AndyScherzinger** [comment]
Exactly,yes :)

**Comment by spacemanspiff2007** [comment]
This would be a huge step foreward since I keep my notes already in a custom folder.
In which version is the available (I still got 1.4.1 from the playstore.

**Comment by mario** [comment]
This will hopefully be in 1.5.0.

**Comment by tobiasKaminsky** [comment]
@mario be careful with ETAs, we do not want to disappoint people and I do not know how much time @AndyScherzinger has is the upcoming weeks ;-)

**Comment by mario** [comment]
@tobiasKaminsky I hope he'll have the time, but if not, I'm also not completely useless :P

**Comment by mario** [comment]
P.S.2. also, I said *hopefully* :D

**Comment by tobiasKaminsky** [comment]
As I understand (and planned with Andy in #285) it should be possible to use any folder across the device with any content as source folder for auto upload, with a nice folder picker. 
Currently the folders are picked when they contain images or videos.
But still it should only be auto upload.

**Comment by mario** [comment]
Understood - thanks for the clarification :)

**Comment by ThomasSteinbach** [comment]
Finally "it should be possible to use any folder across the device" as target for any folder on the Nextcloud.

Example Use Case: I have a password vault file on my Nextcloud under "workspaces/<appname>/vaults/...". Now I want to be able to sync the _vaults_ directory on Nextcloud with the corresponding working directory of my Android app on the smartphones SD card.

Another Use Case could simply be syncing a subfolder of my music library on Nextcloud with the default music directory on my Android device.

**Comment by tobiasKaminsky** [comment]
This is planned for the next steps, see #285. 
First complete 1 way sync: device -> server
After that the other steps ;-)

**Comment by ThomasSteinbach** [comment]
Sorry for the confusion. I didn't meant _sync_. I replaced _sync_ with _pair_ in my previous comment.

I just wanted to highlight that the user might select on both sides, the server and smartphone, an arbitrary folder location for exchanging files. Or is selecting an arbitrary folder on the Nextcloud server a future feature too?

**Comment by tobiasKaminsky** [comment]
As described in #285 currently there is only the way device -> server.
The other way around is planned.
As for the auto upload you can change the target folder on the server, e.g. in which folder on the server you want to store your uploaded images from the camera.

**Comment by codethief** [comment]
@tobiasKaminsky I am really confused right now. Wasn't this issue here about picking an arbitrary folder on the device to correspond to a chosen folder on the server? I don't understand at all what auto-upload functionality has to do with this…

**Comment by tobiasKaminsky** [comment]
@codethief as we do not have a real sync mechanism we can only extend auto upload (on the short term). Again, see #285 as it is described there how we want to proceed.

Maybe I am understanding it wrong, then please write it a little more in detail.

**Comment by leoauri** [comment]
+1, my number one feature request for nextcloud android client :)

**Comment by ianp5a** [comment]
+1. I'm trying to recommend Nextcloud to Android users. Who sync their contacts and calendars. But for many, the Nextcloud app only gets used as a photo backup. And others don't use it at all.

**Comment by mario** [comment]
@AndyScherzinger can we now close this issue?

**Comment by AndyScherzinger** [comment]
Afaik yes, what do you think @tobiasKaminsky

**Comment by mi-hol** [comment]
@tobiasKaminsky seems this question hasn't got a final conclusion.
Actually its the same question a friend asked when I recommended nextcloud instead of buying a NAS,

**Comment by thumos333** [comment]
Hi,

As of now, it seems this functionality has not been implemented. That is there is no proper synchronization both ways on the Android Nextcloud app. Or am I wrong?

It should be something like [Metactrl apps](https://play.google.com/store/search?q=metactrl) do for Dropbox, Google Drive, Mega, etc.: if there are file or folder changes on the server, the app implements them on the Android storage, and vise-versa.

**Comment by AndyScherzinger** [comment]
You are right with both statements: the app doesn't implement this yet and there are other apps out there supporting this for multiple cloud storage products including Nextcloud.

**Comment by thumos333** [comment]
> You are right with both statements: the app doesn't implement this yet and there are other apps out there supporting this for multiple cloud storage products including Nextcloud.

Interestingly, I've just found that when it's about the `Notes` folder, used by the native Notes app, synchronization occurs immediately. Say, when I make changes on my PC, the NC client sends then to the server and in a few seconds I see those changes on my phone. Maybe this functionality has already been implemented in a certain app module, just not yet applied for all synchronized items? I cannot dig into this deeper because I am not a coder.

**Comment by AndyScherzinger** [comment]
My guess is rather in that case that it is a text file which the Android clients opens via a webview on the server itself so it seems to be but isn't. Also if a file exists already and has been downloaded then the Nc client keeps it in sync but (not arbitrary) so we don't sync folders and their subfolders, just files already downloaded.

**Comment by assignUser** [comment]
This ~160$ bug bounty from #19  seems relevant
https://www.bountysource.com/issues/35153835-allow-2-way-sync

**Comment by cukabeka** [comment]
I can second that. Please make this happen, would be extremely helpful to see that.

**Comment by aaronwi** [comment]
7 years later, no progress?

**Comment by JesesePU** [comment]
Claimed this bounty. Wallet: `TU8NBT5iGyMNkLwWmWmgy7tFMbKnafLHcu`

**Comment by JesesePU** [comment]
Claimed this. Wallet: `TU8NBT5iGyMNkLwWmWmgy7tFMbKnafLHcu`

**Comment by JesesePU** [comment]
Claimed this. Wallet: `TU8NBT5iGyMNkLwWmWmgy7tFMbKnafLHcu`

**Comment by JesesePU** [comment]
Claimed this bounty. Wallet: `TU8NBT5iGyMNkLwWmWmgy7tFMbKnafLHcu`
